### DISCUSSION TITLE
New Sites Overview

### DISCUSSION THREAD

**Discussion by brianteeman** [akeeba/panopticon#383] [Show and tell]
I wanted a more dashboard look to the sites overview so I have started to make my own template. Maybe its of interest?

![image](https://github.com/akeeba/panopticon/assets/1296369/f7e29b23-e94b-4a58-a94a-054f53ea5921)

Changes made so far are
- dynamic favicon extracted from the site
- remove the refresh buttons
- improve semantics using definition lists
- group filter moved and now onchange
- search is now dynamic without page refresh - much faster see video at the end
- kept all changes to the template 
- no additional css

![image](https://github.com/akeeba/panopticon/assets/1296369/6a0ef5a9-c20d-4aac-919a-ca450cdf2214)

https://github.com/akeeba/panopticon/assets/1296369/ff21c4ee-35c6-47a6-bb8d-b1d1f43ef8cf

**Comment by nikosdion** [comment]
Can I ask a stupid question? How does it fare if you have 300 sites, or 1000 sites? That's the only reason I am paginating this view and why the search is server-side, not client-side. If you only have a few dozen sites, yeah, client-side search / filtering works. However, that's not something I can ship as you understand. I'd have to fetch filtered information over AJAX and render the interface client-side which for sure is possible, but it starts becoming a pain in the rear for anyone who's trying to make template overrides for customisation.

Regarding favicons, I already have a favicon sniffer in another project. It works better than all the mostly abandoned libraries I have found. I was, indeed, thinking about integrating it with Panopticon at some point.

Regarding the reload buttons, I would still put them there, possibly as items displayed only on-hover / on-focus. They do serve a purpose. Maybe not in your use case, but in the general use case they do.

I could definitely give you an option to have a presentation similar to what you have there. However, before I do that, I'd like to refactor some view templates to move the conditions for presenting certain warnings and information to traits used by the View classes, otherwise things will get _incredibly_ hairy when I add support for WordPress.

**Comment by brianteeman** [comment]
1. tested with 700 sites. It's ok but not great and definitely only really suitable for about 50. Just as the table view is.
2. I tested a few favicon options. This one https://api.faviconkit.com/ seemed to be the most robust and falls back to the initial of the site. I was thinking of some sort of thumbnail snapshot instead for even easier identification
3. I definitely dont want the reload buttons on the dashboard
4. Yes the view templates massively needs to be refactored as there is far too much logic there as well as the joomla specific stuff.
5. This is designed for my use to fit my needs

**Reply by nikosdion** [reply]
I think I am tired and didn't quite explain my question. If search and filtering only applies client-side, how do you manage hundreds of sites? With the table you reload the table, therefore you can “zero into” the sites you are looking for. With the grid view that does not go through the server I assume you'd have to load all sites which is dreadfully slow. Is this the case, or did I miss something important? That's what I meant to ask. Sorry if I am still unclear, it's been a long few days and I didn't have enough caffeine today :D

**Reply by brianteeman** [reply]
no i understood you. its not suitable for a lot of sites.

FYI I was trying to got for something like this https://codepen.io/brianteeman/pen/GRzGXVm where the background changes as well when there is a problem

**Reply by nikosdion** [reply]
This will become easier when I move the error condition detection into the Site model.

I am also thinking about whether it makes sense to go to a more dynamic (read: oodles of JavaScript) approach for handling the site list and the site overview page. However, there are two ways to go about it with pros and cons.

One approach is to use a JS framework (like Vue.js, React.js, …) to render the pages using data retrieved from the server dynamically. This is great for performance, but introduces Yet Another Abstraction Layer which might be pushing it too far for those who are trying to customise the output. I know that Joomla's Media Manager has a dearth of maintainers for similar reasons. Ahem.

The other approach is to load chunks of HTML with XHR/fetch (“AJAX”) and slap them on the page. This has the problem that we need to unregister and re-register JS events, and it's kinda slow as well. On the plus side, you don't have to learn something entirely new and foreign to customise the interface, and the whole thing does not rely on JavaScript.

I hate both alternatives for different reasons. I think I will ultimately have to bite the bullet and go with a JS framework which will probably **NOT** be React, as I've already tried that in the past and it disagrees with me (or, at the very least, its documentation is written in a way which is fundamentally incompatible with my style of learning).

**Reply by brianteeman** [reply]
as I know you are already familiar with it have you considered using datatables for this.

**Reply by nikosdion** [reply]
Are you talking about https://datatables.net or something else? We were using DataTables for jQuery in Akeeba Ticket System up to and including version 4. The problem is that it requires jQuery, and it does not really solve the problem we have at hand. Even if it did, it's jQuery, and jQuery is _slow_. I am staying away from it.

**Reply by brianteeman** [reply]
yes I am talking about that and there is a vanilla version

**Comment by komir2004** [comment]
Hi, IIf I may make a suggestion, maybe to add "backup feature" that shows the status of the last backup and alerts the user if there are any problems.
Something like Yoursites has
![Sites-57-Matching-List-View-My-Joomla-Administration](https://github.com/akeeba/panopticon/assets/17553018/bd6127d5-5fa4-4b3c-80b3-e7ecc3a9ba1e)
Note: This is old version of Yoursites don't know how it is looks now 😥

**Reply by nikosdion** [reply]
Something like that will happen. You can track it at https://github.com/akeeba/panopticon/issues/396

Please note that the backlog is pretty substantial and I am still one person, trying to run a business (need to eat and all that), and maintain this in my spare time. This is something I will implement, I just can't tell you when.

**Comment by brianteeman** [comment]
Another possibility that might be worth considering, especially for instances with a very large number of sutes, would be instead of replacing the current main tabular view would be to introduce a new "dashboard" view as the homepage.

This would be grid style as shown here but **only** for a preselected subset of sites.
This could be based on a dynamic criteria such as sites with an outofdate joomla version or a static criteria such as all sites in a specific group - with a maximum number of sites allowed in that group for performance reasons.

**Reply by nikosdion** [reply]
See https://github.com/akeeba/panopticon/issues/395 and https://github.com/akeeba/panopticon/tree/feature/dashboard

I am using Petite Vue which is great for progressive enhancement, what the dashboard layout is. Here's how it currently looks like.

<img width="1136" alt="Screenshot 2023-12-19 at 00 18 32" src="https://github.com/akeeba/panopticon/assets/256041/5641dbaa-58e1-4554-994d-877ecb72628a">

This can work with arbitrarily large numbers of sites, without taxing the server. Sites are read in chunks of 50, with the default limit being up to 1000 sites displayed at a time. This will be configurable.

**Reply by brianteeman** [reply]
much more readable - thanks - will check out that branch asap to look closer.

not seen Petite Vue before so will have to read up on that as well

**Reply by nikosdion** [reply]
Petite Vue is a subset of Vue.js, by the same author. Combined with a JSON API, it's easy to make basic interactive applications.

**Reply by brianteeman** [reply]
Yeah - I found this interesting explanatory article https://blog.logrocket.com/petite-vue-alpine-alternative-progressive-enhancement/

**Comment by nikosdion** [comment]
@komir2004 This is possible, BUT it will come at a cost. There are three possibilities:

- The data will be completely out of date if the last backup was taken with something other than Panopticon.
- The data will be (mostly) up-to-date, but there needs to be another task running every 15' to update it.
- The data will be (mostly) up-to-date, no need for another task, but the API connector will need to talk directly to Akeeba Backup which means that at some point there will be a breakage.

I am mostly inclined to go for the second option.

@brianteeman There are views with the sites that have outdated Joomla! versions, and the sites that have outdated extensions. There is no need to replicate that in the main page; that's the reason we made separate pages.

I _will_ offer a dashboard view **BUT** it will be dynamically loading 50 to 100 sites at a time (possibly something you can select), eventually presenting everything on your dashboard – maybe it can even be triggered by scrolling to the bottom, we'll see. I was playing with Petite Vue yesterday, as it's minimal (7KiB uncompressed) and _fast_. If I keep adding data to an in-memory JavaScript array I can display as many items as I please to the interface without pagination. Filters would be applied server-side as this allows me to actually transfer far less data per site, while allowing for complex dynamic filters without having to replicate the server-side code in the client-side.

The impetus behind it is not just offer a different layout, but to start applying progressive enhancement throughout the interface. I don't usually say “trust me bro, I have a plan” but I do have one and I do say that in this case 😉

**Reply by brianteeman** [reply]
cool
